### ISSUE TITLE
Random exceptions: Unable to get JavascriptExecutor from CachingRemoteWebElement

### ISSUE THREAD

**Issue by JessefSpecialisterren** [fhoeben/hsac-fitnesse-fixtures#236]
After upgrading the HSAC version of one of our projects, the following exception has been cropping up at seemingly random moments:

```
java.lang.IllegalArgumentException: Unable to get: org.openqa.selenium.JavascriptExecutor from: nl.hsac.fitnesse.fixture.util.selenium.caching.CachingRemoteWebElement
	at nl.hsac.fitnesse.fixture.util.selenium.JavascriptHelper.getJavascriptExecutor(JavascriptHelper.java:34)
	at nl.hsac.fitnesse.fixture.util.selenium.by.BestMatchBy.selectBestElement(BestMatchBy.java:68)
	at nl.hsac.fitnesse.fixture.util.selenium.by.BestMatchBy.findElement(BestMatchBy.java:53)
	at nl.hsac.fitnesse.fixture.util.selenium.by.FirstElementBy.getWebElement(FirstElementBy.java:52)
	at nl.hsac.fitnesse.fixture.util.selenium.by.FirstElementBy.lambda$findElement$1(FirstElementBy.java:44)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1359)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
	at nl.hsac.fitnesse.fixture.util.FirstNonNullHelper.firstNonNull(FirstNonNullHelper.java:98)
	at nl.hsac.fitnesse.fixture.util.FirstNonNullHelper.firstNonNull(FirstNonNullHelper.java:51)
	at nl.hsac.fitnesse.fixture.util.selenium.by.FirstElementBy.findElement(FirstElementBy.java:44)
	at nl.hsac.fitnesse.fixture.util.selenium.by.HeuristicBy.findElement(HeuristicBy.java:39)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.clickSelectOption(BrowserTest.java:682)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.selectFor(BrowserTest.java:659)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at fitnesse.slim.fixtureInteraction.DefaultInteraction.methodInvoke(DefaultInteraction.java:221)
	at nl.hsac.fitnesse.fixture.slim.SlimFixture.invoke(SlimFixture.java:61)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.superInvoke(BrowserTest.java:138)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.lambda$invokedWrappedInWaitUntil$0(BrowserTest.java:104)
	at nl.hsac.fitnesse.fixture.util.selenium.TryAllFramesConditionDecorator.apply(TryAllFramesConditionDecorator.java:33)
	at nl.hsac.fitnesse.fixture.util.selenium.TryAllFramesConditionDecorator.apply(TryAllFramesConditionDecorator.java:17)
	at nl.hsac.fitnesse.fixture.util.selenium.SeleniumHelper.lambda$getConditionIgnoringStaleElement$4(SeleniumHelper.java:902)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:248)
	at nl.hsac.fitnesse.fixture.util.selenium.SeleniumHelper.waitUntil(SeleniumHelper.java:889)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.waitUntilImpl(BrowserTest.java:2008)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.waitUntil(BrowserTest.java:1938)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.invokedWrappedInWaitUntil(BrowserTest.java:131)
	at nl.hsac.fitnesse.fixture.slim.web.BrowserTest.invoke(BrowserTest.java:87)
	at nl.hsac.fitnesse.fixture.slim.SlimFixture.aroundSlimInvoke(SlimFixture.java:41)
	at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at fitnesse.slim.fixtureInteraction.DefaultInteraction.methodInvoke(DefaultInteraction.java:221)
	at fitnesse.slim.fixtureInteraction.DefaultInteraction.callMethod(DefaultInteraction.java:204)
	at fitnesse.slim.fixtureInteraction.DefaultInteraction.invokeMethod(DefaultInteraction.java:188)
	at fitnesse.slim.fixtureInteraction.DefaultInteraction.findAndInvoke(DefaultInteraction.java:166)
	at fitnesse.slim.MethodExecutor.findAndInvoke(MethodExecutor.java:18)
	at fitnesse.slim.FixtureMethodExecutor.execute(FixtureMethodExecutor.java:18)
	at fitnesse.slim.StatementExecutor.getMethodExecutionResult(StatementExecutor.java:139)
	at fitnesse.slim.StatementExecutor.call(StatementExecutor.java:112)
	at fitnesse.slim.instructions.CallInstruction.executeInternal(CallInstruction.java:35)
	at fitnesse.slim.instructions.Instruction.execute(Instruction.java:29)
	at fitnesse.slim.ListExecutor$Executive.executeStatement(ListExecutor.java:49)
	at fitnesse.slim.ListExecutor$Executive.executeStatements(ListExecutor.java:43)
	at fitnesse.slim.ListExecutor.execute(ListExecutor.java:85)
	at fitnesse.slim.SlimServer.executeInstructions(SlimServer.java:82)
	at fitnesse.slim.SlimServer.processOneSetOfInstructions(SlimServer.java:75)
	at fitnesse.slim.SlimServer.tryProcessInstructions(SlimServer.java:62)
	at fitnesse.slim.SlimServer.serve(SlimServer.java:47)
	at fitnesse.slim.SlimService.handle(SlimService.java:169)
	at fitnesse.slim.SlimService.acceptOne(SlimService.java:177)
	at fitnesse.slim.SlimService.accept(SlimService.java:139)
	at fitnesse.slim.SlimService.startWithFactory(SlimService.java:92)
	at fitnesse.slim.SlimService.main(SlimService.java:57)
```

It shows in calls of `|scroll to|`, `|enter|...|as|`, `|click|` and`|select|...|for|` (and possibly more). This project uses Firefox 62.0.3 (64-bits) with geckodriver 0.23.0 (though I recall seeing this exception with 0.22.0), and these errors were produced on a Windows 10 machine.

Is this enough information for you to investigate? If needed, I can try to create an acceptance test that will reproduce the issue, though that may take a while to finish.

**Comment by fhoeben** [comment]
I haven't encountered this issue myself before. Very strange thing to happen randomly. This piece of code is executed for any command after elements are found.

Did you set a search context before this occurs? The context does not seem to be the driver itself, but an element found earlier.
The only thing I can imagine would be that the context element had its `parent` reset for some reason at an earlier point in time. But I just don't know. 

Sorry

**Comment by JessefSpecialisterren** [comment]
It would appear that the problem is fixed in release 3.29.1. I did not do a full regression test, but we have seen the problem with versions 3.27.1 and 3.28.1. Both 3.26.0 and 3.29.1 do not appear to exhibit the problem.

Feel free to close this issue if you like, I'll comment if it surfaces again with the latest release. Or leave it open in case someone with the same issue comes looking, your call 🙂
